test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs

```diff
@@ -1669,7 +1669,153 @@ public async Task CentralTransitiveDependency_Downgrades_IfDowngrades_WereNotRej
+                .DependsOn("D", "[2.0.0]", privateAssets: privateAssets2);
+
+            provider.Package("D", "1.0.0");
+            provider.Package("D", "2.0.0");
 
+            context.LocalLibraryProviders.Add(provider);
+            var walker = new RemoteDependencyWalker(context);
+
+            // Act
+            var rootNode = await DoWalkAsync(walker, "A", framework);
+            var result = rootNode.Analyze();
+
+            // Assert
+            Assert.Equal(expectedConflicts, result.VersionConflicts.Count);
+            Assert.Empty(result.Downgrades);
+
+            if (expectedConflicts == 1)
+            {
+                AssertPath(result.VersionConflicts[0].Conflicting, "A 1.0.0", "B 1.0.0", "D 1.0.0");
+                AssertPath(result.VersionConflicts[0].Selected, "A 1.0.0", "C 1.0.0", "D 2.0.0");
+            }
+        }
+
+        /// <summary>
+        /// A -> B 1.0.0 -> D [1.0.0] (PrivateAssets1)
+        ///              -> C 1.0.0 -> D [2.0.0] (PrivateAssets2)
+        /// </summary>
+        [Theory]
+        [InlineData(null, null, 1)]
+        [InlineData(LibraryIncludeFlags.All, null, 0)]
+        [InlineData(null, LibraryIncludeFlags.All, 0)]
+        [InlineData(LibraryIncludeFlags.All, LibraryIncludeFlags.All, 0)]
+        public async Task PrivateAssetsAll_VersionDowngrades(LibraryIncludeFlags? privateAssets1, LibraryIncludeFlags? privateAssets2, int expectedDowngrades)
+        {
+            var framework = NuGetFramework.Parse("net45");
+
+            var context = new TestRemoteWalkContext();
+            var provider = new DependencyProvider();
+
+            provider.Package("A", "1.0.0")
+                .DependsOn("B", "1.0.0");
+
+            provider.Package("B", "1.0.0")
+                   .DependsOn("C", "1.0.0")
+                   .DependsOn("D", "1.0.0", privateAssets: privateAssets1);
+
+            provider.Package("C", "1.0.0")
+                .DependsOn("D", "2.0.0", privateAssets: privateAssets2);
+
+            provider.Package("D", "1.0.0");
+            provider.Package("D", "2.0.0");
+
+            context.LocalLibraryProviders.Add(provider);
+            var walker = new RemoteDependencyWalker(context);
+
+            // Act
+            var rootNode = await DoWalkAsync(walker, "A", framework);
+            var result = rootNode.Analyze();
+
+            // Assert
+            Assert.Equal(expectedDowngrades, result.Downgrades.Count);
+            Assert.Empty(result.VersionConflicts);
+
+            if (expectedDowngrades == 1)
+            {
+                var downgraded = result.Downgrades[0].DowngradedFrom;
+                var downgradedBy = result.Downgrades[0].DowngradedTo;
+
+                AssertPath(downgraded, "A 1.0.0", "B 1.0.0", "C 1.0.0", "D 2.0.0");
+                AssertPath(downgradedBy, "A 1.0.0", "B 1.0.0", "D 1.0.0");
+            }
+        }
+
+        /// <summary>
+        /// A -> B 1.0.0 (PrivateAssets=?) -> C 2.0.0
+        ///   -> C 1.0.0
```

Comment:
Can you check accuracy of this graph?
`B` version is `2.0.0` not `1.0.0`
Also D is no where here.

Reply:
Thanks for catching this 👍 I've also simplified the graph from the previous test.